Objects/moduleobject.c

```diff
@@ -819,68 +821,104 @@ _Py_module_getattro_impl(PyModuleObject *m, PyObject *name, int suppress)
         return NULL;
     }
-    if (mod_name && PyUnicode_Check(mod_name)) {
-        PyObject *spec;
-        if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__spec__), &spec) < 0) {
+    if (!mod_name || !PyUnicode_Check(mod_name)) {
+        Py_XDECREF(mod_name);
+        PyErr_Format(PyExc_AttributeError,
+                    "module has no attribute '%U'", name);
+        return NULL;
+    }
+    PyObject *spec;
+    if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__spec__), &spec) < 0) {
+        Py_DECREF(mod_name);
+        return NULL;
+    }
+    PyObject *origin = NULL;
+    if (spec) {
+        int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
+        if (rc == -1) {
+            Py_DECREF(spec);
             Py_DECREF(mod_name);
             return NULL;
         }
-        if (suppress != 1) {
-            int rc = _PyModuleSpec_IsInitializing(spec);
-            if (rc > 0) {
-                int valid_spec = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
-                if (valid_spec == -1) {
-                    Py_XDECREF(spec);
-                    Py_DECREF(mod_name);
-                    return NULL;
-                }
-                if (valid_spec == 1 && !PyUnicode_Check(origin)) {
-                    valid_spec = 0;
-                    Py_DECREF(origin);
-                }
-                if (valid_spec == 1) {
-                    PyErr_Format(PyExc_AttributeError,
-                                "partially initialized "
-                                "module '%U' from '%U' has no attribute '%U' "
-                                "(most likely due to a circular import)",
-                                mod_name, origin, name);
-                    Py_DECREF(origin);
-                }
-                else {
-                    PyErr_Format(PyExc_AttributeError,
-                                "partially initialized "
-                                "module '%U' has no attribute '%U' "
-                                "(most likely due to a circular import)",
-                                mod_name, name);
-                }
-            }
-            else if (rc == 0) {
-                rc = _PyModuleSpec_IsUninitializedSubmodule(spec, name);
-                if (rc > 0) {
-                    PyErr_Format(PyExc_AttributeError,
-                                "cannot access submodule '%U' of module '%U' "
-                                "(most likely due to a circular import)",
-                                name, mod_name);
+        if (rc == 1 && !PyUnicode_Check(origin)) {
+            Py_DECREF(origin);
+            origin = NULL;
+        }
+    }
+
+    int is_script_shadowing_stdlib = 0;
+    // Check mod.__name__ in sys.stdlib_module_names
+    // and os.path.dirname(mod.__spec__.origin) == os.getcwd()
+    if (origin) {
+        PyObject *stdlib = PySys_GetObject("stdlib_module_names");
+        if (stdlib && PyAnySet_Check(stdlib) && PySet_Contains(stdlib, mod_name)) {
```

Comment:
`PySet_Contains` can raise an exception.

Reply:
Only if the set isn't a set or key isn't unhashable right? We check for both of these conditions... But let me know and I can add error handling anyway

Reply:
I guess it could be str subclass with broken hash. I'll add the error handling

Reply:
Could you add a test as well? 😄

Reply:
Added a test!